tss-esapi/src/utils/quote.rs

```diff
@@ -0,0 +1,393 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use crate::error::Error;
+use crate::error::Result;
+use crate::WrapperErrorKind;
+use crate::{
+    abstraction::public::AssociatedTpmCurve,
+    interface_types::algorithm::HashingAlgorithm,
+    structures::{
+        Attest, AttestInfo, DigestList, EccSignature, PcrSelectionList, Public, QuoteInfo,
+        Signature,
+    },
+    traits::Marshall,
+};
+use digest::{Digest, DynDigest};
+
+use ecdsa::{
+    hazmat::{DigestPrimitive, VerifyPrimitive},
+    PrimeCurve, SignatureSize, VerifyingKey,
+};
+use elliptic_curve::{
+    generic_array::ArrayLength,
+    point::AffinePoint,
+    sec1::{FromEncodedPoint, ModulusSize, ToEncodedPoint},
+    CurveArithmetic, FieldBytesSize,
+};
+use signature::{hazmat::PrehashVerifier, Verifier};
+
+#[cfg(feature = "rsa")]
+use rsa::{pkcs1v15, pss, RsaPublicKey};
+
+fn verify_ecdsa<C>(
+    public: &Public,
+    message: &[u8],
+    signature: &EccSignature,
+    hashing_algorithm: HashingAlgorithm,
+) -> Result<bool>
+where
+    C: PrimeCurve + CurveArithmetic + DigestPrimitive + AssociatedTpmCurve,
+    AffinePoint<C>: VerifyPrimitive<C> + FromEncodedPoint<C> + ToEncodedPoint<C>,
+    SignatureSize<C>: ArrayLength<u8>,
+    FieldBytesSize<C>: ModulusSize,
+{
+    let Ok(signature) = ecdsa::Signature::<C>::try_from(signature.clone()) else {
+        return Ok(false);
+    };
+    let Ok(public) = elliptic_curve::PublicKey::<C>::try_from(public) else {
+        return Ok(false);
+    };
+
+    let verifying_key = VerifyingKey::from(public);
+
+    match hashing_algorithm {
+        #[cfg(feature = "sha1")]
+        HashingAlgorithm::Sha1 => {
+            let hash = sha1::Sha1::digest(&message);
+            Ok(verifying_key.verify_prehash(&hash, &signature).is_ok())
```

Comment:
This one, I'm actually surprised it works. I was convinced the hash size had to line up with the size of the curve.

Reply:
https://datatracker.ietf.org/doc/html/rfc6979#section-2.3.2

```
   1.  The sequence is first truncated or expanded to length qlen:

       *  if qlen < blen, then the qlen leftmost bits are kept, and
          subsequent bits are discarded;

       *  otherwise, qlen-blen bits (of value zero) are added to the
          left of the sequence (i.e., before the input bits in the
          sequence order)
          
```

TIL. I guess you don't get to sign a sha1 with P521, but this should be taken care of by the verifying library, nothing we should care about here.